File: scala/leetcode_scala/src/main/scala/leetcode/editor/cn/[5]最长回文子串.scala
package leetcode.editor.cn
object LongestPalindromicSubstring {
//leetcode submit region begin(Prohibit modification and deletion)
  object Solution {
    def longestPalindrome(s: String): String = {
      @annotation.tailrec
      def palindrome(str: String, l: Int, r: Int): String = {
        if (l < 0 || r >= str.length || str(l) != str(r))
          str.substring(l + 1, r)
        else palindrome(str, l - 1, r + 1)
      }

      var result = ""
      for { i <- 0 until s.length } {
        val s1 = palindrome(s, i, i)
        val s2 = palindrome(s, i, i + 1)

        if (s1.length > result.length) result = s1
        if (s2.length > result.length) result = s2
      }
      result
    }
  }
//leetcode submit region end(Prohibit modification and deletion)

}

File: dart/leetcode/editor/cn/[5]最长回文子串.dart
//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
  String longestPalindrome(String s) {
    String result = "";
    for (int i = 0; i < s.length; i++) {
      var s1 = _palindrome(s, i, i);
      var s2 = _palindrome(s, i, i + 1);
      if (s1.length > result.length) result = s1;
      if (s2.length > result.length) result = s2;
    }
    return result;
  }

  String _palindrome(String s, int left, int right) {
    if (left < 0 || right >= s.length || s[left] != s[right]) {
      return s.substring(left + 1, right);
    } else {
      return _palindrome(s, left - 1, right + 1);
    }
  }
}
//leetcode submit region end(Prohibit modification and deletion)

void main() {
  print("Hello");
}
